### PR TITLE
Support string type integer for oracle_document_id

### PR DESCRIPTION
**Pull request by cindyyan317**
Seems like rippled supports string-type int for integer field of input json, Clio will starts supporting it.

This PR changed the field oracle_document_id to support both integer type and integer but string-type. For example 1 and "1".

### CODE REVIEW

File: src/rpc/handlers/GetAggregatePrice.hpp [hunk 1]
```diff
@@ -131,23 +132,26 @@ class GetAggregatePriceHandler {
         static auto constexpr ORACLES_MAX = 200;
 
         static auto const oraclesValidator =
-            validation::CustomValidator{[](boost::json::value const& value, std::string_view) -> MaybeError {
+            modifiers::CustomModifier{[](boost::json::value& value, std::string_view) -> MaybeError {
```
Comment (kuznetsss) [inline src/rpc/handlers/GetAggregatePrice.hpp:135]: Why did you remove `const`?

File: src/rpc/common/Modifiers.hpp [hunk 2]
```diff
@@ -100,4 +104,71 @@ struct ToLower final {
...
+    std::function<MaybeError(boost::json::value&, std::string_view)> modifier_;
+
+public:
+    /**
+     * @brief Constructs a custom modifier from any supported callable.
+     *
+     * @tparam Fn The type of callable
+     * @param fn The callable/function object
+     */
+    template <typename Fn>
```
Comment (kuznetsss) [inline src/rpc/common/Modifiers.hpp:155]: Maybe it is better to add a concept for `Fn` to make compile errors nicer.

File: src/rpc/common/Modifiers.hpp [hunk 3]
```diff
@@ -100,4 +104,71 @@ struct ToLower final {
...
+     * @brief Modify the JSON value according to the custom modifier function stored.
+     *
+     * @param value The JSON value representing the outer object
+     * @param key The key used to retrieve the tested value from the outer object
+     * @return Any compatible user-provided error if modify/verify failed; otherwise no error is returned
+     */
+    [[nodiscard]] MaybeError
+    modify(boost::json::value& value, std::string_view key) const
+    {
+        if (not value.is_object() or not value.as_object().contains(key.data()))
```
Comment (kuznetsss) [inline src/rpc/common/Modifiers.hpp]: I don't think you need `.data()` here and in other places

File: src/rpc/common/MetaProcessors.hpp [hunk 4]
```diff
@@ -189,26 +194,36 @@ class WithCustomError final {
     [[nodiscard]] MaybeError
     verify(boost::json::value const& value, std::string_view key) const
     {
-        if (auto const res = requirement.verify(value, key); not res)
-            return Error{error};
-
+        if constexpr (SomeRequirement<decltype(reqOrModifier)>) {
```
Comment (kuznetsss) [inline src/rpc/common/MetaProcessors.hpp]: I think you could do
```cpp
requires SomeRequirement<SomeRequirementOrModifier>
```
then you wouldn't need `unsupported_v`.

File: src/rpc/common/MetaProcessors.hpp [hunk 5]
```diff
@@ -160,22 +163,24 @@ class IfType final {
 };
 
 /**
- * @brief A meta-processor that wraps a validator and produces a custom error in case the wrapped validator fails.
+ * @brief A meta-processor that wraps a validator or modifier and produces a custom error in case the wrapped validator
+ * or modifier fails.
  */
-template <typename SomeRequirement>
+template <typename SomeRequirementOrModifier>
```
Comment (kuznetsss) [inline src/rpc/common/MetaProcessors.hpp]: Maybe better `ReuirementOrModifierType`? Because `Some...` we usually use for a concept names.

And probably better to add a concept here that it is either requirement or modifier.

File: src/rpc/common/MetaProcessors.hpp [hunk 6]
```diff
@@ -189,26 +194,36 @@ class WithCustomError final {
...
     verify(boost::json::value& value, std::string_view key) const
     {
-        if (auto const res = requirement.verify(value, key); not res)
-            return Error{error};
-
+        if constexpr (SomeRequirement<decltype(reqOrModifier)>) {
+            if (auto const res = reqOrModifier.verify(value, key); not res)
+                return Error{error};
+        } else if constexpr (SomeModifier<decltype(reqOrModifier)>) {
+            if (auto const res = reqOrModifier.modify(value, key); not res)
```
Comment (kuznetsss) [inline src/rpc/common/MetaProcessors.hpp]: It is a bit weird that method `verify()` performs a modification. Is it not possible to separate verification and modification?

File: tests/unit/rpc/BaseTests.cpp [hunk 7]
```diff
@@ -604,3 +604,55 @@ TEST_F(RPCBaseTest, ToLowerModifier)
...
 }
+
+TEST_F(RPCBaseTest, ToNumberModifier)
+{
+    auto spec = RpcSpec{
+        {"str", ToNumber{}},
+    };
+
+    auto passingInput = json::parse(R"({ "str": [] })");
+    ASSERT_TRUE(spec.process(passingInput));
```
Comment (kuznetsss) [inline tests/unit/rpc/BaseTests.cpp:616]: I think, it is better to check that `passingInput` didn't change after this.

Also probably better to split this test in parametrised tests, because currently `passingInput` is shared and multiple tests cases could affect each other.

File: tests/unit/rpc/BaseTests.cpp [hunk 8]
```diff
@@ -604,3 +604,55 @@ TEST_F(RPCBaseTest, ToLowerModifier)
...
+
+    passingInput = json::parse(R"({ "str2": "TesT" })");
+    ASSERT_TRUE(spec.process(passingInput));
+
+    passingInput = json::parse(R"([])");
+    ASSERT_TRUE(spec.process(passingInput));
+
+    passingInput = json::parse(R"({ "str": "123" })");
```
Comment (kuznetsss) [inline tests/unit/rpc/BaseTests.cpp:624]: What if it would be `{ "str": "123.456" }`?

File: src/rpc/common/MetaProcessors.hpp (hunk 6 above)
Comment (cindyyan317) [reply src/rpc/common/MetaProcessors.hpp]: Yes. We can discuss how to improve this. I believe it will need framework change.

File: tests/unit/rpc/BaseTests.cpp [hunk 9]
```diff
@@ -604,3 +604,58 @@ TEST_F(RPCBaseTest, ToLowerModifier)
...
+
+TEST_F(RPCBaseTest, CustomModifier)
+{
+    auto const customModifier = CustomModifier{[](json::value& value, std::string_view /* key */) -> MaybeError {
+        if (value.is_string()) {
+            value = json::value("modified");
+            return MaybeError{};
+        }
+        return Error{rpc::Status{"Uh oh"}};
+    }};
```
Comment (kuznetsss) [inline tests/unit/rpc/BaseTests.cpp]: I think a better way to test it is to use a mock function:
```cpp
testing::StrictMock<testing::MockFunction<MaybeError(json::value& value, std::string_view)>> mockModifier;
auto const customModifier = CustomModifier{mockModifier.AsStdFunction()};
```

and then you will be able to check whether the mock was called when you call `spec.process()`.

**Review by kuznetsss — APPROVED** [review]
👍
